components/brave_wallet/browser/cardano/cardano_transaction.cc

```diff
@@ -0,0 +1,489 @@
+  return true;
+}
+
+template <size_t SZ>
+bool ReadHexByteArrayTo(const base::Value::Dict& dict,
+                        std::string_view key,
+                        std::array<uint8_t, SZ>& to) {
+  auto* str = dict.FindString(key);
+  if (!str) {
+    return false;
+  }
+  return base::HexStringToSpan(*str, to);
+}
+
+}  // namespace
+
+CardanoTransaction::CardanoTransaction() = default;
+CardanoTransaction::~CardanoTransaction() = default;
+CardanoTransaction::CardanoTransaction(const CardanoTransaction& other) =
+    default;
+CardanoTransaction& CardanoTransaction::operator=(
+    const CardanoTransaction& other) = default;
+CardanoTransaction::CardanoTransaction(CardanoTransaction&& other) = default;
+CardanoTransaction& CardanoTransaction::operator=(CardanoTransaction&& other) =
+    default;
+bool CardanoTransaction::operator==(const CardanoTransaction& other) const =
+    default;
+bool CardanoTransaction::operator!=(const CardanoTransaction& other) const =
+    default;
+
+CardanoTransaction::Outpoint::Outpoint() = default;
+CardanoTransaction::Outpoint::~Outpoint() = default;
+CardanoTransaction::Outpoint::Outpoint(const Outpoint& other) = default;
+CardanoTransaction::Outpoint& CardanoTransaction::Outpoint::operator=(
+    const Outpoint& other) = default;
+CardanoTransaction::Outpoint::Outpoint(Outpoint&& other) = default;
+CardanoTransaction::Outpoint& CardanoTransaction::Outpoint::operator=(
+    Outpoint&& other) = default;
+
+base::Value::Dict CardanoTransaction::Outpoint::ToValue() const {
+  base::Value::Dict dict;
+
+  dict.Set("txid", base::HexEncode(txid));
+  dict.Set("index", static_cast<int>(index));
+
+  return dict;
+}
+
+// static
+std::optional<CardanoTransaction::Outpoint>
+CardanoTransaction::Outpoint::FromValue(const base::Value::Dict& value) {
+  Outpoint result;
+
+  auto* txid_hex = value.FindString("txid");
+  if (!txid_hex) {
+    return std::nullopt;
+  }
+  if (!base::HexStringToSpan(*txid_hex, result.txid)) {
+    return std::nullopt;
+  }
+
+  auto index_value = value.FindInt("index");
+  if (!index_value) {
+    return std::nullopt;
+  }
+  result.index = *index_value;
+
+  return result;
+}
+
+CardanoTransaction::TxInput::TxInput() = default;
+CardanoTransaction::TxInput::~TxInput() = default;
+CardanoTransaction::TxInput::TxInput(const CardanoTransaction::TxInput& other) =
+    default;
+CardanoTransaction::TxInput& CardanoTransaction::TxInput::operator=(
+    const CardanoTransaction::TxInput& other) = default;
+CardanoTransaction::TxInput::TxInput(CardanoTransaction::TxInput&& other) =
+    default;
+CardanoTransaction::TxInput& CardanoTransaction::TxInput::operator=(
+    CardanoTransaction::TxInput&& other) = default;
+
+base::Value::Dict CardanoTransaction::TxInput::ToValue() const {
+  base::Value::Dict dict;
+
+  dict.Set("utxo_address", utxo_address.ToString());
+  dict.Set("utxo_outpoint", utxo_outpoint.ToValue());
+  dict.Set("utxo_value", base::NumberToString(utxo_value));
```

Comment:
All these types have some of this `base::Value` machinery for fields that we would be getting for free if we were using idl types to handle these fields. I feel a lot of these structs should be generated with the json schema compiler, and then we should just have helper functions to do conversions to and from our types, but I wonder if it would be better if we didn't mix the schema and the concrete data.

Reply:
Idls are in most cases enough when we need to parse external json into a c++ object.
But when we have a c++ class in the first place and need to persist it - idls are not practical when used directly.

Reply:
The JSON schema compiler is for compiling schema into C++ structures so we avoid handling each and every field with custom code. Whenever there's a couple of types where `FromValue`/`ToValue` pairs occur, those types most likely can be mechanically generated with the compiler. It looks like the underlying data here has a `base::Value::Dict` representation, so I think we would benefit having representation in the compiler, so we don't have to be writing it ourselves.

Reply:
Again, idl is a schema-first thing. 
I'd love to see an example of idl scheme which would produce a c++ struct like 
```
  struct TxInput {
...
    CardanoAddress utxo_address;
    Outpoint utxo_outpoint;
    uint64_t utxo_value = 0;
  };
```
which would be convenient to use in business logic.

Reply:
I particularly would like to see us exploring the use of IDLs to generate these types. That doesn't mean `TxInput` or anything has to change. It would only mean that there's another set of types that mirror the JSON structure. I'm saying this because all this boilerplate would be gone. The issue with boilerplate like this is that we have to read it, when we visit the code, and it is also easier to make mistakes.

Reply:
Well I see now what you mean. Most likely there always would be intermediate `TxInputGenerated` struct hiding `base::Value` conversions from `TxInput`.
Such approach would have both pros(less boilerplate, etc) and cons(extra conversion step, idl types are quite limited, etc). Needs some experimenting whether it scales well.

Reply:
I couldn't find any TODOs, issues in the code about this thread? What resolution are we looking for?

Reply:
That's even better, nice.